theseus/geometry/se3.py

```diff
@@ -0,0 +1,167 @@
+        ret = torch.zeros(self.shape[0], 3, 4).to(dtype=self.dtype, device=self.device)
+        ret[:, :, :3] = self.data[:, :3, :3].transpose(1, 2)
+        ret[:, :, 3] = -(ret[:, :3, :3] @ self.data[:, :3, 3].unsqueeze(2)).view(-1, 3)
+        return SE3(data=ret)
+
+    def to_matrix(self) -> torch.Tensor:
+        ret = torch.zeros(self.shape[0], 4, 4).to(dtype=self.dtype, device=self.device)
+        ret[:, :3] = self.data
+        ret[:, 3, 3] = 1
+        return ret
+
+    def update_from_x_y_z_quaternion(self, x_y_z_quaternion: torch.Tensor):
+        if x_y_z_quaternion.ndim != 2 and x_y_z_quaternion.shape[1] != 7:
+            raise ValueError("x_y_z_quaternion can only be 7-D vectors.")
+
+        batch_size = x_y_z_quaternion.shape[0]
+        self.data = torch.empty(batch_size, 3, 4).to(
+            device=x_y_z_quaternion.device, dtype=x_y_z_quaternion.dtype
+        )
+        self[:, :3, :3] = SO3.unit_quaternion_to_matrix(x_y_z_quaternion[:, 3:])
+        self[:, :3, 3] = x_y_z_quaternion[:, :3]
+
+    def update_from_rot_and_trans(self, rotation: SO3, translation: Point3):
+        if rotation.shape[0] != translation.shape[0]:
+            raise ValueError("rotation and translation must have the same size.")
+
+        if rotation.dtype != translation.dtype:
+            raise ValueError("rotation and translation must be of the same type.")
+
+        if rotation.device != translation.device:
+            raise ValueError("rotation and translation must be on the same device.")
+
+        self.data = torch.cat((rotation.data, translation.data.unsqueeze(2)), dim=2)
+
+    @staticmethod
+    def hat(tangent_vector: torch.Tensor) -> torch.Tensor:
+        _check = tangent_vector.ndim == 2 and tangent_vector.shape[1] == 6
+        if not _check:
+            raise ValueError("Invalid vee matrix for SE(3).")
+        matrix = torch.zeros(tangent_vector.shape[0], 4, 4).to(
+            dtype=tangent_vector.dtype, device=tangent_vector.device
+        )
+        matrix[:, :3, :3] = SO3.hat(tangent_vector[:, 3:])
+        matrix[:, :3, 3] = tangent_vector[:, :3]
+
+        return matrix
+
+    @staticmethod
+    def vee(matrix: torch.Tensor) -> torch.Tensor:
+        SE3._hat_matrix_check(matrix)
+        return torch.cat((matrix[:, :3, 3], SO3.vee(matrix[:, :3, :3])), dim=1)
+
+    def _transform_shape_check(self, point: Union[Point3, torch.Tensor]):
+        raise NotImplementedError
+
+    def _copy_impl(self, new_name: Optional[str] = None) -> "SE3":
+        return SE3(data=self.data.clone(), name=new_name)
+
+    # only added to avoid casting downstream
+    def copy(self, new_name: Optional[str] = None) -> "SE3":
+        return cast(SE3, super().copy(new_name=new_name))
+
+    def transform_to(
+        self,
+        point: Union[Point3, torch.Tensor],
+        jacobians: Optional[List[torch.Tensor]] = None,
+    ) -> Point3:
+        raise NotImplementedError
+
+    def untransform_to(
```

Comment:
Do you mean `transform_from` or is this different?

Reply:
Yeah, I mean `transform_from` and will change the name. 

Also, it seems that no `transform_from` is defined in [`SE2`](https://github.com/facebookresearch/theseus/blob/taoshaf.add_se3/theseus/geometry/se2.py). Should we add it to `SE2`.

Reply:
Created issue #69 for this to address in a separate PR. I'd say this is lower priority, so it can wait until the 3D groups are finished.